storage/src/object_storage/file_slice_stream.rs

```diff
@@ -0,0 +1,118 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::io::{self, SeekFrom};
+use std::ops::Range;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use bytes::Bytes;
+use futures::{Stream, StreamExt};
+use tokio::io::{AsyncRead, AsyncSeek, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+/// Offers a stream of byte over a specific range of bytes in a file.
+///
+/// This struct is useful when uploading an object to S3.
+#[derive(Debug)]
+pub struct FileSliceStream<R> {
+    inner: ReaderStream<R>,
+    remaining: u64,
+}
+
+impl<R> FileSliceStream<R>
+where
+    R: AsyncRead + AsyncSeek + Unpin,
+{
+    pub async fn try_new(mut reader: R, range: Range<u64>) -> io::Result<Self> {
+        assert!(range.end >= range.start);
```

Comment:
Do we want to allow requesting a fetch of 0 bytes? `range.end >= range.start`
Could possibly be useful for checking object reachability but is this really the intent?
we could also early return if this the case avoiding seeks

Reply:
Good point! Let's return an Error if range.end < range.start. It is much better

Fetch 0 bytes is supported but no need to optimize it.